src/goto-programs/goto_contractor.cpp

```diff
@@ -0,0 +1,430 @@
+//
+// Created by Mohannad Aldughaim on 09/01/2022.
+//
+
+#include "goto_contractor.h"
+
+void goto_contractort::get_constraints(goto_functionst goto_functions)
+{
+  auto function = goto_functions.function_map.find("c:@F@main");
+  auto it = function->second.body.instructions.begin();
+  while(it != function->second.body.instructions.end())
+  {
+    if(it->is_assert())
+      constraint = create_constraint_from_expr2t(it->guard);
+
+    it++;
+  }
+}
+
+void goto_contractort::get_intervals(goto_functionst goto_functions)
+{
+  //only main here
+  auto function = goto_functions.function_map.find("c:@F@main");
+  auto it = function->second.body.instructions.begin();
+
+  while(it != function->second.body.instructions.end())
+  {
+    if(it->is_assume())
+      parse_intervals(it->guard);
+
+    it++;
+  }
+}
+
+void goto_contractort::parse_intervals(irep_container<expr2t> expr)
+{
+  symbol2tc symbol;
+  long value;
+
+  if(!is_comp_expr(expr))
+    return;
+
+  std::shared_ptr<relation_data> rel;
+  rel = std::dynamic_pointer_cast<relation_data>(expr);
+
+  //side1 1 is a symbol or typecast to symbol
+  auto side1 = rel->side_1;
+  //side 2 is always a number.
+  auto side2 = rel->side_2;
+
+  auto obj = get_base_object(side1);
+  if(is_symbol2t(obj))
+    symbol = to_symbol2t(obj);
+  else
+    return;
+  ///keeping this for later
+  //  if(is_typecast2t(side1))
+  //    if(is_symbol2t(to_typecast2t(side1).from))
+  //      symbol = to_symbol2t(to_typecast2t(side1).from);
+  //    else
+  //      return;
+  //  else if(is_symbol2t(side1))
+  //    symbol = to_symbol2t(side1);
+  //  else
+  //    return;
+
+  if(!is_constant_int2t(side2) && !is_constant_floatbv2t(side2))
+    return;
+  if(is_constant_int2t(side2))
+    value = to_constant_int2t(side2).as_long();
+  else
+    return;
+
+  int index = map->find(symbol->get_symbol_name());
+  if(index == -1)
+    return;
+  switch(expr->expr_id)
+  {
+  case expr2t::expr_ids::greaterthan_id:
+  case expr2t::expr_ids::greaterthanequal_id:
+    map->update_lb_interval(value, index);
+    break;
+  case expr2t::expr_ids::lessthan_id:
+  case expr2t::expr_ids::lessthanequal_id:
+    map->update_ub_interval(value, index);
+    break;
+  case expr2t::expr_ids::equality_id:
+    map->update_lb_interval(value, index);
+    map->update_ub_interval(value, index);
+    break;
+  default:;
+    //error
+  }
+}
+
+void goto_contractort::insert_assume(
+  goto_functionst goto_functions,
+  IntervalVector new_intervals)
+{
+  loopst loop;
+  for(auto &function_loop : function_loops)
+    loop = function_loop;
+
+  auto loop_exit = loop.get_original_loop_exit();
+
+  goto_programt dest(message_handler);
+
+  auto it = goto_functions.function_map.find("c:@F@main");
+  auto goto_function = it->second;
+
+  for(int i = 0; i < map->size(); i++)
```

Comment:
should `size_t` be used here instead of `int`?

Reply:
changed.